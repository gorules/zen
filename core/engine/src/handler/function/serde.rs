use crate::handler::function::error::ResultExt;
use itertools::Itertools;
use rquickjs::{Ctx, FromJs, IntoAtom, IntoJs, Type, Value as QValue};
use serde_json::{json, Map, Number, Value};

#[derive(Debug)]
pub(crate) struct JsValue(pub(crate) Value);

impl<'js> FromJs<'js> for JsValue {
    fn from_js(ctx: &Ctx<'js>, v: QValue<'js>) -> rquickjs::Result<Self> {
        let computed_value = match v.type_of() {
            Type::Uninitialized | Type::Undefined | Type::Null => Value::Null,
            Type::Bool => Value::Bool(v.as_bool().or_throw_msg(ctx, "failed to convert to bool")?),
            Type::Int => Value::Number(Number::from(
                v.as_int().or_throw_msg(ctx, "failed to convert to int")?,
            )),
            Type::BigInt => Value::Number(Number::from(
                v.as_big_int()
                    .map(|b| b.clone().to_i64().ok())
                    .flatten()
                    .or_throw_msg(ctx, "failed to convert to number")?,
            )),
            Type::Float => Value::Number(
                v.as_float()
                    .map(|n| Number::from_f64(n))
                    .flatten()
                    .or_throw_msg(ctx, "failed to convert to number")?,
            ),
            Type::String => Value::String(
                v.as_string()
                    .map(|s| s.to_string().ok())
                    .flatten()
                    .or_throw_msg(ctx, "failed to convert to string")?,
            ),
            Type::Array => {
                let arr = v
                    .as_array()
                    .or_throw_msg(ctx, "failed to convert to array")?;

                let js_arr: Vec<Value> = arr
                    .iter::<QValue>()
                    .map_ok(|n| JsValue::from_js(ctx, n.clone()).map(|js_val| js_val.0))
                    .flatten()
                    .try_collect()
                    .or_throw(ctx)?;

                Value::Array(js_arr)
            }
            Type::Object => {
                let object = v
                    .as_object()
                    .or_throw_msg(ctx, "failed to convert to object")?;

                let js_object: Map<String, Value> = object
                    .props::<String, QValue>()
                    .map_ok(|(key, value)| {
                        JsValue::from_js(ctx, value.clone()).map(|js_val| (key, js_val.0))
                    })
                    .flatten()
                    .try_collect()
                    .or_throw(ctx)?;

                Value::Object(js_object)
            }
            Type::Exception => {
                let exception = v
                    .as_exception()
                    .or_throw_msg(ctx, "failed to convert to exception")?;

                let message = exception.message().unwrap_or_default();
                let description = exception.to_string();

                json!({ "message": message, "description": description })
            }
            Type::Function => json!("[Function]"),
            Type::Module => json!("[Module]"),
            Type::Constructor => json!("[Constructor]"),
            Type::Symbol => json!("[Symbol]"),
            Type::Unknown => json!("[Unknown]"),
            Type::Promise => {
                let promise = v.as_promise().or_throw(ctx)?;
                let val: JsValue = promise.finish()?;
                val.0
            }
        };

        Ok(JsValue(computed_value))
    }
}

impl<'js> IntoJs<'js> for JsValue {
    fn into_js(self, ctx: &Ctx<'js>) -> rquickjs::Result<QValue<'js>> {
        let res = match self.0 {
            Value::Null => QValue::new_null(ctx.clone()),
            Value::Bool(b) => QValue::new_bool(ctx.clone(), b),
            Value::Number(n) => QValue::new_number(
                ctx.clone(),
                n.as_f64()
                    .or_throw_msg(ctx, "failed to convert float to number")?,
            ),
            Value::String(str) => str.into_js(ctx)?,
            Value::Array(arr) => {
                let qarr = rquickjs::Array::new(ctx.clone())?;
                for (idx, item) in arr.into_iter().enumerate() {
                    qarr.set(idx, JsValue(item))?;
                }

                qarr.into_value()
            }
            Value::Object(map) => {
                let qmap = rquickjs::Object::new(ctx.clone())?;
                for (key, value) in map.into_iter() {
                    qmap.set(key.into_atom(ctx)?, JsValue(value))?;
                }

                qmap.into_value()
            }
        };

        Ok(res)
    }
}
