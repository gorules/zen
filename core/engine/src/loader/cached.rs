use std::collections::HashMap;
use std::future::Future;
use std::sync::Arc;

use tokio::sync::Mutex;

use crate::loader::{DecisionLoader, LoaderResponse};
use crate::model::DecisionContent;

pub struct CachedLoader<Loader: DecisionLoader + 'static> {
    loader: Arc<Loader>,
    cache: Mutex<HashMap<String, Arc<DecisionContent>>>,
}

impl<Loader: DecisionLoader + 'static> From<Arc<Loader>> for CachedLoader<Loader> {
    fn from(value: Arc<Loader>) -> Self {
        Self {
            loader: value,
            cache: Mutex::new(HashMap::new()),
        }
    }
}

impl<Loader: DecisionLoader + 'static> DecisionLoader for CachedLoader<Loader> {
    fn load<'a>(&'a self, key: &'a str) -> impl Future<Output = LoaderResponse> + 'a {
        async move {
            let mut cache = self.cache.lock().await;
            if let Some(content) = cache.get(key) {
                return Ok(content.clone());
            }

            let decision_content = self.loader.load(key).await?;
            cache.insert(key.to_string(), decision_content.clone());
            Ok(decision_content)
        }
    }
}
