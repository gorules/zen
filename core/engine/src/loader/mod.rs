mod closure;
mod filesystem;
mod memory;
mod noop;

pub use closure::ClosureLoader;
pub use filesystem::{FilesystemLoader, FilesystemLoaderOptions};
pub use memory::MemoryLoader;
pub use noop::NoopLoader;

use async_trait::async_trait;

use crate::model::DecisionContent;
use std::fmt::Debug;
use std::sync::Arc;
use thiserror::Error;

pub type LoaderResult<T> = Result<T, Box<LoaderError>>;
pub type LoaderResponse = LoaderResult<Arc<DecisionContent>>;

/// Trait used for implementing a loader for decisions
#[async_trait]
pub trait DecisionLoader {
    async fn load(&self, key: &str) -> LoaderResponse;
}

#[derive(Error, Debug)]
pub enum LoaderError {
    #[error("Loader did not find item with key {0}")]
    NotFound(String),
    #[error("Loader failed internally on key {key}: {source}.")]
    Internal {
        key: String,
        #[source]
        source: anyhow::Error,
    },
}
