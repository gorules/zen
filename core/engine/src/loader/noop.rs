use crate::loader::{DecisionLoader, LoaderError, LoaderResponse};
use anyhow::anyhow;
use async_trait::async_trait;

/// Default loader which always fails
#[derive(Default, Debug)]
pub struct NoopLoader;

#[async_trait]
impl DecisionLoader for NoopLoader {
    async fn load(&self, key: &str) -> LoaderResponse {
        Err(LoaderError::Internal {
            key: key.to_string(),
            source: anyhow!("Loader is no-op"),
        }
        .into())
    }
}
