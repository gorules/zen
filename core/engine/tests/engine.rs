use std::io::Read;
use std::ops::Deref;
use std::path::Path;
use std::sync::Arc;

use serde::Deserialize;
use serde_json::json;

use zen_engine::loader::{LoaderError, MemoryLoader};
use zen_engine::model::DecisionNodeKind;
use zen_engine::{DecisionEngine, EvaluationError, EvaluationOptions};

use crate::support::{create_fs_loader, load_raw_test_data, load_test_data};

mod support;

#[tokio::test]
#[cfg_attr(miri, ignore)]
async fn engine_memory_loader() {
    let memory_loader = Arc::new(MemoryLoader::default());
    memory_loader.add("table", load_test_data("table.json"));
    memory_loader.add("function", load_test_data("function.json"));

    let engine = DecisionEngine::new_arc(memory_loader.clone());
    let table = engine.evaluate("table", &json!({ "input": 12 })).await;
    let function = engine.evaluate("function", &json!({ "input": 12 })).await;

    memory_loader.remove("function");
    let not_found = engine.evaluate("function", &json!({})).await;

    assert_eq!(table.unwrap().result, json!({"output": 10}));
    assert_eq!(function.unwrap().result, json!({"output": 24}));
    assert_eq!(not_found.unwrap_err().to_string(), "Loader error");
}

#[tokio::test]
#[cfg_attr(miri, ignore)]
async fn engine_filesystem_loader() {
    let engine = DecisionEngine::new(create_fs_loader());
    let table = engine.evaluate("table.json", &json!({ "input": 12 })).await;
    let function = engine
        .evaluate("function.json", &json!({ "input": 12 }))
        .await;
    let not_found = engine.evaluate("invalid_file", &json!({})).await;

    assert_eq!(table.unwrap().result, json!({"output": 10}));
    assert_eq!(function.unwrap().result, json!({"output": 24}));
    assert_eq!(not_found.unwrap_err().to_string(), "Loader error");
}

#[tokio::test]
#[cfg_attr(miri, ignore)]
async fn engine_closure_loader() {
    let engine = DecisionEngine::async_loader(|key| {
        // TODO: Improve once async closures become stable in Rust
        let mv_key = key.to_string();

        async move {
            match mv_key.as_str() {
                "function" => Ok(Arc::new(load_test_data("function.json"))),
                "table" => Ok(Arc::new(load_test_data("table.json"))),
                _ => Err(LoaderError::NotFound(mv_key).into()),
            }
        }
    });

    let table = engine.evaluate("table", &json!({ "input": 12 })).await;
    let function = engine.evaluate("function", &json!({ "input": 12 })).await;
    let not_found = engine.evaluate("invalid_file", &json!({})).await;

    assert_eq!(table.unwrap().result, json!({"output": 10}));
    assert_eq!(function.unwrap().result, json!({"output": 24}));
    assert_eq!(not_found.unwrap_err().to_string(), "Loader error");
}

#[tokio::test]
async fn engine_noop_loader() {
    // Default engine is noop
    let engine = DecisionEngine::default();
    let result = engine.evaluate("any.json", &json!({})).await;

    assert_eq!(result.unwrap_err().to_string(), "Loader error");
}

#[tokio::test]
async fn engine_get_decision() {
    let engine = DecisionEngine::new(create_fs_loader());

    assert!(engine.get_decision("table.json").await.is_ok());
    assert!(engine.get_decision("any.json").await.is_err());
}

#[tokio::test]
async fn engine_create_decision() {
    let engine = DecisionEngine::default();
    engine.create_decision(load_test_data("table.json").into());
}

#[tokio::test]
#[cfg_attr(miri, ignore)]
async fn engine_errors() {
    let engine = DecisionEngine::new(create_fs_loader());

    let infinite_fn = engine.evaluate("infinite-function.json", &json!({})).await;
    match infinite_fn.unwrap_err().deref() {
        EvaluationError::NodeError(e) => {
            assert_eq!(e.node_id, "e0fd96d0-44dc-4f0e-b825-06e56b442d78");
            assert_eq!(e.source.to_string(), "Timeout exceeded");
        }
        _ => assert!(false, "Wrong error type"),
    }

    let recursive = engine.evaluate("recursive-table1.json", &json!({})).await;
    match recursive.unwrap_err().deref() {
        EvaluationError::NodeError(e) => {
            assert_eq!(e.source.to_string(), "Depth limit exceeded")
        }
        _ => assert!(false, "Depth limit not exceeded"),
    }
}

#[tokio::test]
async fn engine_with_trace() {
    let engine = DecisionEngine::new(create_fs_loader());

    let table_r = engine.evaluate("table.json", &json!({ "input": 12 })).await;
    let table_opt_r = engine
        .evaluate_with_opts(
            "table.json",
            &json!({ "input": 12 }),
            EvaluationOptions {
                trace: Some(true),
                max_depth: None,
            },
        )
        .await;

    let table = table_r.unwrap();
    let table_opt = table_opt_r.unwrap();

    assert!(table.trace.is_none());
    assert!(table_opt.trace.is_some());

    let trace = table_opt.trace.unwrap();
    assert_eq!(trace.len(), 3); // trace for each node
}

#[tokio::test]
#[cfg_attr(miri, ignore)]
async fn engine_function_imports() {
    let mut function_content = load_test_data("function.json");

    let imports_js_path = Path::new("js").join("imports.js");
    let mut replace_buffer = load_raw_test_data(imports_js_path.to_str().unwrap());
    let mut replace_data = String::new();
    replace_buffer.read_to_string(&mut replace_data).unwrap();

    function_content.nodes.iter_mut().for_each(|node| {
        if let DecisionNodeKind::FunctionNode { content } = &mut node.kind {
            let _ = std::mem::replace(content, replace_data.clone());
        }
    });

    let decision = DecisionEngine::default().create_decision(function_content.into());
    let response = decision.evaluate(&json!({})).await.unwrap();

    #[derive(Deserialize, Debug)]
    #[serde(rename_all = "camelCase")]
    struct GraphResult {
        bigjs_tests: Vec<bool>,
        bigjs_valid: bool,
        dayjs_valid: bool,
        moment_valid: bool,
    }

    let result = serde_json::from_value::<GraphResult>(response.result).unwrap();

    assert!(result.bigjs_tests.iter().all(|v| *v));
    assert!(result.bigjs_valid);
    assert!(result.dayjs_valid);
    assert!(result.moment_valid);
}
