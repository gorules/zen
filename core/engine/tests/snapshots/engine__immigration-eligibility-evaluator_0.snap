---
source: core/engine/tests/engine.rs
expression: serialized_result
---
performance: "[perf]"
result:
  applicant:
    currentResidence: France
    dateOfBirth: 1985-05-15
    fullName: Jane Smith
    nationality: United Kingdom
  assessment:
    additionalFactorsReason: In-demand skills
    additionalFactorsScore:
      "$serde_json::private::Number": "80"
    backgroundReason: Background check clear
    backgroundScore:
      "$serde_json::private::Number": "100"
    backgroundStatus: pending
    documentationReason: Documentation complete
    documentationScore:
      "$serde_json::private::Number": "100"
    documentationStatus: pending
    educationReason: Advanced education
    educationScore:
      "$serde_json::private::Number": "80"
    experienceReason: Significant work experience
    experienceScore:
      "$serde_json::private::Number": "80"
    finalReason: Application meets criteria
    finalStatus: approved
    languageReason: Advanced language proficiency
    languageScore:
      "$serde_json::private::Number": "75"
    overallScore:
      "$serde_json::private::Number": "85.83333333333333333333333333"
  background:
    criminalHistory: false
    immigrationViolations: false
    previousVisaRejections:
      "$serde_json::private::Number": "0"
    securityConcerns: false
  documentation:
    medicalCheckComplete: true
    passportValid: true
    proofOfFundsProvided: true
    supportingDocuments:
      - Reference Letter
      - Property Deed
      - Bank Statements
    visaApplicationComplete: true
  qualifications:
    educationLevel: master
    familyConnections: false
    languageProficiency: advanced
    skillsInDemand: true
    workExperienceYears:
      "$serde_json::private::Number": "8"
trace:
  additionalFactorsTable:
    id: additionalFactorsTable
    input:
      applicant:
        currentResidence: France
        dateOfBirth: 1985-05-15
        fullName: Jane Smith
        nationality: United Kingdom
      assessment:
        backgroundReason: Background check clear
        backgroundScore:
          "$serde_json::private::Number": "100"
        backgroundStatus: pending
        documentationReason: Documentation complete
        documentationScore:
          "$serde_json::private::Number": "100"
        documentationStatus: pending
        educationReason: Advanced education
        educationScore:
          "$serde_json::private::Number": "80"
        experienceReason: Significant work experience
        experienceScore:
          "$serde_json::private::Number": "80"
        languageReason: Advanced language proficiency
        languageScore:
          "$serde_json::private::Number": "75"
      background:
        criminalHistory: false
        immigrationViolations: false
        previousVisaRejections:
          "$serde_json::private::Number": "0"
        securityConcerns: false
      documentation:
        medicalCheckComplete: true
        passportValid: true
        proofOfFundsProvided: true
        supportingDocuments:
          - Reference Letter
          - Property Deed
          - Bank Statements
        visaApplicationComplete: true
      qualifications:
        educationLevel: master
        familyConnections: false
        languageProficiency: advanced
        skillsInDemand: true
        workExperienceYears:
          "$serde_json::private::Number": "8"
    name: assessAdditionalFactors
    order:
      "$serde_json::private::Number": "6"
    output:
      applicant:
        currentResidence: France
        dateOfBirth: 1985-05-15
        fullName: Jane Smith
        nationality: United Kingdom
      assessment:
        additionalFactorsReason: In-demand skills
        additionalFactorsScore:
          "$serde_json::private::Number": "80"
        backgroundReason: Background check clear
        backgroundScore:
          "$serde_json::private::Number": "100"
        backgroundStatus: pending
        documentationReason: Documentation complete
        documentationScore:
          "$serde_json::private::Number": "100"
        documentationStatus: pending
        educationReason: Advanced education
        educationScore:
          "$serde_json::private::Number": "80"
        experienceReason: Significant work experience
        experienceScore:
          "$serde_json::private::Number": "80"
        languageReason: Advanced language proficiency
        languageScore:
          "$serde_json::private::Number": "75"
      background:
        criminalHistory: false
        immigrationViolations: false
        previousVisaRejections:
          "$serde_json::private::Number": "0"
        securityConcerns: false
      documentation:
        medicalCheckComplete: true
        passportValid: true
        proofOfFundsProvided: true
        supportingDocuments:
          - Reference Letter
          - Property Deed
          - Bank Statements
        visaApplicationComplete: true
      qualifications:
        educationLevel: master
        familyConnections: false
        languageProficiency: advanced
        skillsInDemand: true
        workExperienceYears:
          "$serde_json::private::Number": "8"
    performance: "[perf]"
    traceData:
      index:
        "$serde_json::private::Number": "1"
      reference_map:
        qualifications.familyConnections: false
        qualifications.skillsInDemand: true
      rule:
        _id: rule2
        "qualifications.familyConnections[i6-2]": "false"
        "qualifications.skillsInDemand[i6-1]": "true"
  backgroundCheck:
    id: backgroundCheck
    input:
      applicant:
        currentResidence: France
        dateOfBirth: 1985-05-15
        fullName: Jane Smith
        nationality: United Kingdom
      assessment:
        documentationReason: Documentation complete
        documentationScore:
          "$serde_json::private::Number": "100"
        documentationStatus: pending
      background:
        criminalHistory: false
        immigrationViolations: false
        previousVisaRejections:
          "$serde_json::private::Number": "0"
        securityConcerns: false
      documentation:
        medicalCheckComplete: true
        passportValid: true
        proofOfFundsProvided: true
        supportingDocuments:
          - Reference Letter
          - Property Deed
          - Bank Statements
        visaApplicationComplete: true
      qualifications:
        educationLevel: master
        familyConnections: false
        languageProficiency: advanced
        skillsInDemand: true
        workExperienceYears:
          "$serde_json::private::Number": "8"
    name: evaluateBackground
    order:
      "$serde_json::private::Number": "2"
    output:
      applicant:
        currentResidence: France
        dateOfBirth: 1985-05-15
        fullName: Jane Smith
        nationality: United Kingdom
      assessment:
        backgroundReason: Background check clear
        backgroundScore:
          "$serde_json::private::Number": "100"
        backgroundStatus: pending
        documentationReason: Documentation complete
        documentationScore:
          "$serde_json::private::Number": "100"
        documentationStatus: pending
      background:
        criminalHistory: false
        immigrationViolations: false
        previousVisaRejections:
          "$serde_json::private::Number": "0"
        securityConcerns: false
      documentation:
        medicalCheckComplete: true
        passportValid: true
        proofOfFundsProvided: true
        supportingDocuments:
          - Reference Letter
          - Property Deed
          - Bank Statements
        visaApplicationComplete: true
      qualifications:
        educationLevel: master
        familyConnections: false
        languageProficiency: advanced
        skillsInDemand: true
        workExperienceYears:
          "$serde_json::private::Number": "8"
    performance: "[perf]"
    traceData:
      index:
        "$serde_json::private::Number": "5"
      reference_map:
        background.criminalHistory: false
        background.immigrationViolations: false
        background.previousVisaRejections:
          "$serde_json::private::Number": "0"
        background.securityConcerns: false
      rule:
        _id: rule6
        "background.criminalHistory[i2-1]": "false"
        "background.immigrationViolations[i2-2]": "false"
        "background.previousVisaRejections[i2-4]": "0"
        "background.securityConcerns[i2-3]": "false"
  documentationCheck:
    id: documentationCheck
    input:
      applicant:
        currentResidence: France
        dateOfBirth: 1985-05-15
        fullName: Jane Smith
        nationality: United Kingdom
      background:
        criminalHistory: false
        immigrationViolations: false
        previousVisaRejections:
          "$serde_json::private::Number": "0"
        securityConcerns: false
      documentation:
        medicalCheckComplete: true
        passportValid: true
        proofOfFundsProvided: true
        supportingDocuments:
          - Reference Letter
          - Property Deed
          - Bank Statements
        visaApplicationComplete: true
      qualifications:
        educationLevel: master
        familyConnections: false
        languageProficiency: advanced
        skillsInDemand: true
        workExperienceYears:
          "$serde_json::private::Number": "8"
    name: verifyDocumentation
    order:
      "$serde_json::private::Number": "1"
    output:
      applicant:
        currentResidence: France
        dateOfBirth: 1985-05-15
        fullName: Jane Smith
        nationality: United Kingdom
      assessment:
        documentationReason: Documentation complete
        documentationScore:
          "$serde_json::private::Number": "100"
        documentationStatus: pending
      background:
        criminalHistory: false
        immigrationViolations: false
        previousVisaRejections:
          "$serde_json::private::Number": "0"
        securityConcerns: false
      documentation:
        medicalCheckComplete: true
        passportValid: true
        proofOfFundsProvided: true
        supportingDocuments:
          - Reference Letter
          - Property Deed
          - Bank Statements
        visaApplicationComplete: true
      qualifications:
        educationLevel: master
        familyConnections: false
        languageProficiency: advanced
        skillsInDemand: true
        workExperienceYears:
          "$serde_json::private::Number": "8"
    performance: "[perf]"
    traceData:
      index:
        "$serde_json::private::Number": "4"
      reference_map:
        documentation.medicalCheckComplete: true
        documentation.passportValid: true
        documentation.proofOfFundsProvided: true
        documentation.visaApplicationComplete: true
      rule:
        _id: rule5
        "documentation.medicalCheckComplete[i1-4]": "true"
        "documentation.passportValid[i1-1]": "true"
        "documentation.proofOfFundsProvided[i1-3]": "true"
        "documentation.visaApplicationComplete[i1-2]": "true"
  educationTable:
    id: educationTable
    input:
      applicant:
        currentResidence: France
        dateOfBirth: 1985-05-15
        fullName: Jane Smith
        nationality: United Kingdom
      assessment:
        backgroundReason: Background check clear
        backgroundScore:
          "$serde_json::private::Number": "100"
        backgroundStatus: pending
        documentationReason: Documentation complete
        documentationScore:
          "$serde_json::private::Number": "100"
        documentationStatus: pending
      background:
        criminalHistory: false
        immigrationViolations: false
        previousVisaRejections:
          "$serde_json::private::Number": "0"
        securityConcerns: false
      documentation:
        medicalCheckComplete: true
        passportValid: true
        proofOfFundsProvided: true
        supportingDocuments:
          - Reference Letter
          - Property Deed
          - Bank Statements
        visaApplicationComplete: true
      qualifications:
        educationLevel: master
        familyConnections: false
        languageProficiency: advanced
        skillsInDemand: true
        workExperienceYears:
          "$serde_json::private::Number": "8"
    name: assessEducation
    order:
      "$serde_json::private::Number": "3"
    output:
      applicant:
        currentResidence: France
        dateOfBirth: 1985-05-15
        fullName: Jane Smith
        nationality: United Kingdom
      assessment:
        backgroundReason: Background check clear
        backgroundScore:
          "$serde_json::private::Number": "100"
        backgroundStatus: pending
        documentationReason: Documentation complete
        documentationScore:
          "$serde_json::private::Number": "100"
        documentationStatus: pending
        educationReason: Advanced education
        educationScore:
          "$serde_json::private::Number": "80"
      background:
        criminalHistory: false
        immigrationViolations: false
        previousVisaRejections:
          "$serde_json::private::Number": "0"
        securityConcerns: false
      documentation:
        medicalCheckComplete: true
        passportValid: true
        proofOfFundsProvided: true
        supportingDocuments:
          - Reference Letter
          - Property Deed
          - Bank Statements
        visaApplicationComplete: true
      qualifications:
        educationLevel: master
        familyConnections: false
        languageProficiency: advanced
        skillsInDemand: true
        workExperienceYears:
          "$serde_json::private::Number": "8"
    performance: "[perf]"
    traceData:
      index:
        "$serde_json::private::Number": "1"
      reference_map:
        qualifications.educationLevel: master
      rule:
        _id: rule2
        "qualifications.educationLevel[i3-1]": "\"master\""
  experienceTable:
    id: experienceTable
    input:
      applicant:
        currentResidence: France
        dateOfBirth: 1985-05-15
        fullName: Jane Smith
        nationality: United Kingdom
      assessment:
        backgroundReason: Background check clear
        backgroundScore:
          "$serde_json::private::Number": "100"
        backgroundStatus: pending
        documentationReason: Documentation complete
        documentationScore:
          "$serde_json::private::Number": "100"
        documentationStatus: pending
        educationReason: Advanced education
        educationScore:
          "$serde_json::private::Number": "80"
      background:
        criminalHistory: false
        immigrationViolations: false
        previousVisaRejections:
          "$serde_json::private::Number": "0"
        securityConcerns: false
      documentation:
        medicalCheckComplete: true
        passportValid: true
        proofOfFundsProvided: true
        supportingDocuments:
          - Reference Letter
          - Property Deed
          - Bank Statements
        visaApplicationComplete: true
      qualifications:
        educationLevel: master
        familyConnections: false
        languageProficiency: advanced
        skillsInDemand: true
        workExperienceYears:
          "$serde_json::private::Number": "8"
    name: assessWorkExperience
    order:
      "$serde_json::private::Number": "4"
    output:
      applicant:
        currentResidence: France
        dateOfBirth: 1985-05-15
        fullName: Jane Smith
        nationality: United Kingdom
      assessment:
        backgroundReason: Background check clear
        backgroundScore:
          "$serde_json::private::Number": "100"
        backgroundStatus: pending
        documentationReason: Documentation complete
        documentationScore:
          "$serde_json::private::Number": "100"
        documentationStatus: pending
        educationReason: Advanced education
        educationScore:
          "$serde_json::private::Number": "80"
        experienceReason: Significant work experience
        experienceScore:
          "$serde_json::private::Number": "80"
      background:
        criminalHistory: false
        immigrationViolations: false
        previousVisaRejections:
          "$serde_json::private::Number": "0"
        securityConcerns: false
      documentation:
        medicalCheckComplete: true
        passportValid: true
        proofOfFundsProvided: true
        supportingDocuments:
          - Reference Letter
          - Property Deed
          - Bank Statements
        visaApplicationComplete: true
      qualifications:
        educationLevel: master
        familyConnections: false
        languageProficiency: advanced
        skillsInDemand: true
        workExperienceYears:
          "$serde_json::private::Number": "8"
    performance: "[perf]"
    traceData:
      index:
        "$serde_json::private::Number": "1"
      reference_map:
        qualifications.workExperienceYears:
          "$serde_json::private::Number": "8"
      rule:
        _id: rule2
        "qualifications.workExperienceYears[i4-1]": ">= 7"
  input:
    id: input
    input: ~
    name: immigrationApplication
    order:
      "$serde_json::private::Number": "0"
    output:
      applicant:
        currentResidence: France
        dateOfBirth: 1985-05-15
        fullName: Jane Smith
        nationality: United Kingdom
      background:
        criminalHistory: false
        immigrationViolations: false
        previousVisaRejections:
          "$serde_json::private::Number": "0"
        securityConcerns: false
      documentation:
        medicalCheckComplete: true
        passportValid: true
        proofOfFundsProvided: true
        supportingDocuments:
          - Reference Letter
          - Property Deed
          - Bank Statements
        visaApplicationComplete: true
      qualifications:
        educationLevel: master
        familyConnections: false
        languageProficiency: advanced
        skillsInDemand: true
        workExperienceYears:
          "$serde_json::private::Number": "8"
    performance: "[perf]"
    traceData: ~
  languageTable:
    id: languageTable
    input:
      applicant:
        currentResidence: France
        dateOfBirth: 1985-05-15
        fullName: Jane Smith
        nationality: United Kingdom
      assessment:
        backgroundReason: Background check clear
        backgroundScore:
          "$serde_json::private::Number": "100"
        backgroundStatus: pending
        documentationReason: Documentation complete
        documentationScore:
          "$serde_json::private::Number": "100"
        documentationStatus: pending
        educationReason: Advanced education
        educationScore:
          "$serde_json::private::Number": "80"
        experienceReason: Significant work experience
        experienceScore:
          "$serde_json::private::Number": "80"
      background:
        criminalHistory: false
        immigrationViolations: false
        previousVisaRejections:
          "$serde_json::private::Number": "0"
        securityConcerns: false
      documentation:
        medicalCheckComplete: true
        passportValid: true
        proofOfFundsProvided: true
        supportingDocuments:
          - Reference Letter
          - Property Deed
          - Bank Statements
        visaApplicationComplete: true
      qualifications:
        educationLevel: master
        familyConnections: false
        languageProficiency: advanced
        skillsInDemand: true
        workExperienceYears:
          "$serde_json::private::Number": "8"
    name: assessLanguage
    order:
      "$serde_json::private::Number": "5"
    output:
      applicant:
        currentResidence: France
        dateOfBirth: 1985-05-15
        fullName: Jane Smith
        nationality: United Kingdom
      assessment:
        backgroundReason: Background check clear
        backgroundScore:
          "$serde_json::private::Number": "100"
        backgroundStatus: pending
        documentationReason: Documentation complete
        documentationScore:
          "$serde_json::private::Number": "100"
        documentationStatus: pending
        educationReason: Advanced education
        educationScore:
          "$serde_json::private::Number": "80"
        experienceReason: Significant work experience
        experienceScore:
          "$serde_json::private::Number": "80"
        languageReason: Advanced language proficiency
        languageScore:
          "$serde_json::private::Number": "75"
      background:
        criminalHistory: false
        immigrationViolations: false
        previousVisaRejections:
          "$serde_json::private::Number": "0"
        securityConcerns: false
      documentation:
        medicalCheckComplete: true
        passportValid: true
        proofOfFundsProvided: true
        supportingDocuments:
          - Reference Letter
          - Property Deed
          - Bank Statements
        visaApplicationComplete: true
      qualifications:
        educationLevel: master
        familyConnections: false
        languageProficiency: advanced
        skillsInDemand: true
        workExperienceYears:
          "$serde_json::private::Number": "8"
    performance: "[perf]"
    traceData:
      index:
        "$serde_json::private::Number": "1"
      reference_map:
        qualifications.languageProficiency: advanced
      rule:
        _id: rule2
        "qualifications.languageProficiency[i5-1]": "\"advanced\""
  overallAssessment:
    id: overallAssessment
    input:
      applicant:
        currentResidence: France
        dateOfBirth: 1985-05-15
        fullName: Jane Smith
        nationality: United Kingdom
      assessment:
        additionalFactorsReason: In-demand skills
        additionalFactorsScore:
          "$serde_json::private::Number": "80"
        backgroundReason: Background check clear
        backgroundScore:
          "$serde_json::private::Number": "100"
        backgroundStatus: pending
        documentationReason: Documentation complete
        documentationScore:
          "$serde_json::private::Number": "100"
        documentationStatus: pending
        educationReason: Advanced education
        educationScore:
          "$serde_json::private::Number": "80"
        experienceReason: Significant work experience
        experienceScore:
          "$serde_json::private::Number": "80"
        languageReason: Advanced language proficiency
        languageScore:
          "$serde_json::private::Number": "75"
      background:
        criminalHistory: false
        immigrationViolations: false
        previousVisaRejections:
          "$serde_json::private::Number": "0"
        securityConcerns: false
      documentation:
        medicalCheckComplete: true
        passportValid: true
        proofOfFundsProvided: true
        supportingDocuments:
          - Reference Letter
          - Property Deed
          - Bank Statements
        visaApplicationComplete: true
      qualifications:
        educationLevel: master
        familyConnections: false
        languageProficiency: advanced
        skillsInDemand: true
        workExperienceYears:
          "$serde_json::private::Number": "8"
    name: finalAssessment
    order:
      "$serde_json::private::Number": "7"
    output:
      applicant:
        currentResidence: France
        dateOfBirth: 1985-05-15
        fullName: Jane Smith
        nationality: United Kingdom
      assessment:
        additionalFactorsReason: In-demand skills
        additionalFactorsScore:
          "$serde_json::private::Number": "80"
        backgroundReason: Background check clear
        backgroundScore:
          "$serde_json::private::Number": "100"
        backgroundStatus: pending
        documentationReason: Documentation complete
        documentationScore:
          "$serde_json::private::Number": "100"
        documentationStatus: pending
        educationReason: Advanced education
        educationScore:
          "$serde_json::private::Number": "80"
        experienceReason: Significant work experience
        experienceScore:
          "$serde_json::private::Number": "80"
        finalReason: Application meets criteria
        finalStatus: approved
        languageReason: Advanced language proficiency
        languageScore:
          "$serde_json::private::Number": "75"
        overallScore:
          "$serde_json::private::Number": "85.83333333333333333333333333"
      background:
        criminalHistory: false
        immigrationViolations: false
        previousVisaRejections:
          "$serde_json::private::Number": "0"
        securityConcerns: false
      documentation:
        medicalCheckComplete: true
        passportValid: true
        proofOfFundsProvided: true
        supportingDocuments:
          - Reference Letter
          - Property Deed
          - Bank Statements
        visaApplicationComplete: true
      qualifications:
        educationLevel: master
        familyConnections: false
        languageProficiency: advanced
        skillsInDemand: true
        workExperienceYears:
          "$serde_json::private::Number": "8"
    performance: "[perf]"
    traceData:
      index:
        "$serde_json::private::Number": "4"
      reference_map:
        (assessment.documentationScore + assessment.backgroundScore + assessment.educationScore + assessment.experienceScore + assessment.languageScore + assessment.additionalFactorsScore) / 6:
          "$serde_json::private::Number": "85.83333333333333333333333333"
        assessment.backgroundStatus: pending
        assessment.documentationStatus: pending
      rule:
        "(assessment.documentationScore + assessment.backgroundScore + assessment.educationScore + assessment.experienceScore + assessment.languageScore + assessment.additionalFactorsScore) / 6[i7-3]": ">= 70"
        _id: rule5
        "assessment.backgroundStatus[i7-2]": ""
        "assessment.documentationStatus[i7-1]": ""
